daisy_workflows/image_import/enterprise_linux/translate.py

```diff
@@ -87,16 +88,75 @@
 '''
 
 
+class NonFatalErrors:
+  """Non-fatal errors that occur during translation.
+
+  Not all errors are fatal; when a non-fatal error occurs, track it
+  here. If translation fails, these can be used to create a customized
+  error message.
+  """
+  def __init__(self):
+    # When `yum` isn't found, either the wrong --os was specified, or
+    # the disk contains EL but wasn't mountable by GuestOS.
+    self.yum_command_not_found = False
+
+    # BYOL was specified, but an active subscription wasn't found.
+    self.subscription_manager_status_error = False
+
+    # YUM will fail if any of its repositories are unreachable. Typically
+    # this occurs when there's a local repo that isn't available during
+    # translate. Think: DVD, intranet, NFS.
+    self.yum_unreachable_repo = False
+
+  def __repr__(self):
+    return str(self.__dict__)
+
+
+def check_repos(g: guestfs.GuestFS, errs: NonFatalErrors):
+  """Check for unreachable repos.
+
+  YUM fails if any of its repos are unreachable. Running `yum updateinfo`
+  will have a non-zero return code when it fail to update any of its repos.
+  """
+  v = 'yum updateinfo -v'
+  p = run(g, v)
+  logging.debug('yum updateinfo -v: {}'.format(p))
+  if p.code != 0:
+    errs.yum_unreachable_repo = True
+
+
+def check_yum_on_path(g: guestfs.GuestFS, errs: NonFatalErrors):
+  """Check whether the `yum` command is available.
+
+  If `yum` isn't found, errs is updated.
+  """
+  p = run(g, 'yum --version')
+  logging.debug('yum --version: {}'.format(p))
+  if p.code != 0:
+    errs.yum_command_not_found = True
+
+
+def check_rhel_license(g: guestfs.GuestFS, errs: NonFatalErrors):
+  """Check for an active RHEL license.
+
+  If a license isn't found, errs is updated.
+  """
+  p = run(g, 'subscription-manager status')
+  logging.debug('subscription-manager: {}'.format(p))
+  if p.code != 0:
+    errs.subscription_manager_status_error = True
+
+
 def upgrade_ca_certificates(g: guestfs.GuestFS):
   """Upgrade ca-certificates, if it is installed.
 
   Stale or missing CA certificates can cause yum operations to fail.
   """
 
-  try:
-    g.sh('rpm -V ca-certificates')
-  except RuntimeError as e:
-    logging.debug('ca-certificates not found. Skipping upgrade.', e)
+  p = run(g, 'rpm -qa ca-certificates | grep ca-certificates')
+  if p.code != 0:
+    logging.debug(
+        'ca-certificates not found. Skipping upgrade. {}'.format(p))
```

Comment:
What's the difference between -V and -qa? Can exception still occur with this command?

Reply:
`-V` checks that the package is installed, and then performs validations (think: checksums, file permissions). This caused false negatives when the package was installed, but failed one of the validations.

That said, your comment prompted me to dig some more, and I created separate PR that improves the performance of the ca-certificates update: https://github.com/GoogleCloudPlatform/compute-image-tools/pull/1423